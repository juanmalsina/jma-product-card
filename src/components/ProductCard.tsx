import React from "react";
import { createContext } from "react";
import { useProduct } from "../hooks/useProducts";
import {
  ProductCardProps,
  ProductContextProps,
} from "../interfaces/interfaces";
import styles from "../styles/styles.module.css";

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: ProductCardProps) => {
  const { counter, increaseBy, maxCount, isMaxCountReached, reset } =
    useProduct({
      onChange,
      product,
      value,
      initialValues,
    });
  return (
    <Provider
      value={{
        counter,
        increaseBy,
        maxCount,
        product,
      }}
    >
      <div style={style} className={`${styles.productCard} ${className}`}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: initialValues?.maxCount,
          product,

          reset,
          increaseBy,
        })}
      </div>
    </Provider>
  );
};