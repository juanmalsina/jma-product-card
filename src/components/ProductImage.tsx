import React from 'react';
import { useContext } from 'react';
import { ProductContext } from './ProductCard';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';

interface ProductImageProps {
  img?: string;
  className?: string;
  style?: React.CSSProperties;
}

export const ProductImage = ({ img, className, style }: ProductImageProps) => {
  const { product } = useContext(ProductContext);
  const imageToShow = product?.img || img || noImage;
  return (
    <img
      style={style}
      className={`${styles.productImg} ${className}`}
      src={imageToShow}
      alt="coffe mug"
    />
  );
};
