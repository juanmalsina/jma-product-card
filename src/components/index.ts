import { ProducTitle } from './ProducTitle';
import { ProductButtons } from './ProductButtons';
import { ProductImage } from './ProductImage';
import { ProductCard } from './ProductCard';

export { ProducTitle, ProductButtons, ProductImage, ProductCard };
