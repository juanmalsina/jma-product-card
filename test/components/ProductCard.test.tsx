import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { ProductCard } from '../../src/components';
import { product2 } from '../../src/mocks/data';

describe('ProducCard', () => {
  test('debe mostrar el componente correctamente con un children personalizado', () => {
    const wrapper = renderer.create(
      <ProductCard
        initialValues={{ maxCount: 10, count: 0 }}
        product={product2}
      >
        {() => <h1> Product Card </h1>}
      </ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('debe incrementar el contador', () => {
    const wrapper = renderer.create(
      <ProductCard
        initialValues={{ maxCount: 10, count: 0 }}
        product={product2}
      >
        {({ count, increaseBy }) => (
          <>
            <h1> Product Card </h1>
            <span>{count}</span>
            <button onClick={() => increaseBy(1)}>+1</button>
          </>
        )}
      </ProductCard>
    );

    let tree = wrapper.toJSON();
    expect(tree).toMatchSnapshot();
    act(() => {
      (tree as any).children[2].props.onClick();
    });
    tree = wrapper.toJSON();

    expect((tree as any).children[1].children[0]).toBe('1');
  });
});
