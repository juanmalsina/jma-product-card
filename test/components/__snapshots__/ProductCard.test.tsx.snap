// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProducCard debe incrementar el contador 1`] = `
<div
  className="productCard undefined"
>
  <h1>
     Product Card 
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  >
    +1
  </button>
</div>
`;

exports[`ProducCard debe mostrar el componente correctamente con un children personalizado 1`] = `
<div
  className="productCard undefined"
>
  <h1>
     Product Card 
  </h1>
</div>
`;
